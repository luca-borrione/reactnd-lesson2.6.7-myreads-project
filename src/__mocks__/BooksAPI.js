import * as data from './allBooks.json'
import { TShelfKey } from '../types'

const DELAY = 10;

const getDelay = () => DELAY * (Math.floor(Math.random() * 10) + 1); // random between 1 - 10

export const MOCKED = true;

let allBooks;
export const __resetBooks = () => {
	allBooks = JSON.parse(JSON.stringify(data.books)); // cloning
};
__resetBooks();

const getBooksInShelves = () =>
	allBooks.filter( ({ shelf }) => shelf !== TShelfKey.NONE );


export const getAll = () => {
	console.log('==> there getAll');
	const booksInShelves = getBooksInShelves();

	return new Promise( resolve => {
		const delay = getDelay();
		console.log('delay: '+delay);

		setTimeout(() => {
			resolve(booksInShelves);
		}, delay);

	});
};

export const get = bookId => {
	console.log('==> there get');
	const book = allBooks.find( ({ id }) => id === bookId );

	return new Promise( resolve => {
		const delay = getDelay();
		console.log('delay: '+delay);

		setTimeout(() => {
			resolve(book);
		}, delay);

	});
};

const formatUpdatedBooks = (books) => {
	const fBooks = {};

	Object.keys(TShelfKey).forEach( key => {
		const shelfKey = TShelfKey[key];
		if (shelfKey !== TShelfKey.NONE) {
			fBooks[shelfKey] = books
							.filter( ({ shelf }) => shelf === shelfKey )
							.map( book => book.id );
		}
	});

	return fBooks
};

export const update = (book, shelf) => {
	console.log('==> there update');
	const existingBook = allBooks.find( ({ id }) => id === book.id );
	existingBook.shelf = shelf;

	const booksInShelves = getBooksInShelves();

	return new Promise( resolve => {
		const delay = getDelay();
		console.log('delay: '+delay);

		setTimeout(() => {
			resolve(formatUpdatedBooks(booksInShelves));
		}, delay);

	});

};

export const search = (query, maxResults) => {
	console.log('==> there search');
	const re = new RegExp(`^${query}`, 'i');
	const result = allBooks.filter( ({ title, authors }) => {
		return re.test(title);
	})
	.slice(0, maxResults);

	return new Promise( resolve => {
		const delay = getDelay();
		console.log('delay: '+delay);

		setTimeout(() => {
			resolve(result);
		}, delay);

	});

	// console.log(result);
	// existingBook.shelf = shelf;

	// const booksInShelves = getBooksInShelves();

	// return new Promise( resolve => {
	// 	const delay = getDelay();
	// 	console.log('delay: '+delay);

	// 	setTimeout(() => {
	// 		resolve(formatUpdatedBooks(booksInShelves));
	// 	}, delay);

	// });

};


export const search1 = (query, maxResults) =>
  fetch(`${api}/search`, {
    method: 'POST',
    headers: {
      ...headers,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ query, maxResults })
  }).then(res => res.json())
    .then(data => data.books)

