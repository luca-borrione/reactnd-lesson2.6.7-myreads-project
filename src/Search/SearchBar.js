import React from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';
import * as BooksAPI from '../BooksAPI';

/**
 * @module
 * @name SearchBar
 * @extends React.Component
 * @description
 * Shows an input type field with whih the user can perform a search for new books.
 */
class SearchBar extends React.Component {
  /**
   * @member
   * @name SEARCH_LIMIT
   * @description the search is limited to this amount of elements
   * @constant {number}
   * @default 20
   * @static
   * @private
   */
  static SEARCH_LIMIT = 20;


  /**
   * @member
   * @name DEBOUNCING_TIME
   * @description only one search allowed within the frame of this amount milliseconds
   * @constant {number}
   * @default 250
   * @static
   * @private
   */
  static DEBOUNCING_TIME = 250; // milliseconds


  /**
   * @member
   * @name BOOKS_STATUS
   * @description Collection of possible status of the books retrieved by a search
   * @property {string} READY - The search has been performed successfully
   * @property {string} ERROR - An error occurred while performing the search
   * @static
   * @private
   */
  static BOOKS_STATUS = {
    READY: 'BOOKS_STATUS.READY',
    ERROR: 'BOOKS_STATUS.ERROR',
  };


  /**
   * @name propTypes
   * @description Intended types of the props passed to the component
   * @property {function} showResult - see [showResult]{@link module:SearchPage~showResult}
   * @static
   */
  static propTypes = {
    showResult: PropTypes.func.isRequired,
  };


  /**
   * @name state
   * @property {string} query - search query typed by the user in the search bar
   * @static
   * @private
   */
  state = {
    query: '',
  };


  /**
   * @member
   * @name async
   * @description
   * Collection of promises useful when making unit tests.
   * @property {Promise} onTyping - waiting for the books to be fetched
   *                                when the user types in the search bar
   */
  async = {
    onTyping: null,
  };

  constructor(props) {
    super(props);
    const { DEBOUNCING_TIME } = this.constructor;

    this.onTyping = this.onTyping.bind(this);
    this.search = debounce(this.search, DEBOUNCING_TIME);
  }

  /**
   * @method
   * @name shouldComponentUpdate
   * @param {Object} nextProps - ignored
   * @param {state} nextState - refers to [state]{@link module:SearchBar~sstate}
   * @returns {boolean} - false if the query hasn't changed
   * @private
   */
  shouldComponentUpdate(nextProps, nextState) {
    const { query } = this.state;
    return query !== nextState.query;
  }

  /**
   * @method
   * @name onTyping
   * @description
   * Updates the internal state according to the query typed by the user,
   * then it clears the books list if the textfield is empty,
   * otherwise it performs a search based on the typed query.
   * @param {SyntheticEvent} event
   * @returns {void}
   * @private
   */
  onTyping(event) {
    const query = event.target.value;
    const { showResult } = this.props;
    this.async.onTyping = new Promise(async (resolve) => {
      await this.setState({ query });
      const books = await this.fetchBooks(query);
      await showResult(books);
      resolve(books);
    });
  }

  /**
   * @method
   * @name search
   * @description
   * Promise executor that performs a search to find books matching
   * the query string typed by the user.
   * If no books are found, or if the search completes with an error,
   * then an empty array is returned so that the list under the searchbar is cleared.
   * The promise is rejected if an exception is triggered while performng the search.
   * @param {string} query - query string typed by the user
   * @param {function} [resolve] - optional callback triggered when the promise is fullfilled
   * @param {function} [reject] - optional callback triggered when the promise
   * @returns {TBook[]}
   * @private
   */
  async search(query, resolve, reject) {
    const { SEARCH_LIMIT } = this.constructor;
    const { getBookShelf } = this.props;

    let books;

    if (query === '') {
      books = [];
      return resolve(books);
    }

    try {
      const fetchedBooks = await BooksAPI.search(query, SEARCH_LIMIT);
      if (!fetchedBooks || fetchedBooks.error) {
        books = [];
      } else {
        books = fetchedBooks.map((book) => {
          // The book literal objects returned by the api search
          // don't contain the shelf property, so we need to define it by ourselves.
          // The ones returned by the mock api search alredy contain it instead.
          const shelf = book.shelf || getBookShelf(book.id);
          return { ...book, shelf };
        });
      }
      return resolve(books);
    } catch (error) {
      return reject(error);
    }
  }


  /**
   * @method
   * @name fetchBooks
   * @description
   * Fetches the books according to the query string typed by the user,
   * triggering the search itself and then applying to the books array
   * an internal status READY if the search completes correctly
   * or an internal statue ERROR if an something went wrong.
   * @param {string} query - query string typed by the user
   * @return {Promise.<TBook[]>}
   * @private
   */
  fetchBooks(query) {
    const { BOOKS_STATUS } = this.constructor;
    return new Promise((resolve, reject) => {
      this.search(query, resolve, reject);
    }).then((found) => {
      const books = [...found];
      books.status = BOOKS_STATUS.READY;
      return books;
    }).catch((error) => {
      if (/testing/.test(error.message) === false) {
        console.error('ERROR: ', error); // eslint-disable-line no-console
      }
      const books = [];
      books.status = BOOKS_STATUS.ERROR;
      return books;
    });
  }


  /**
   * @method
   * @name render
   * @description
   * Displays an input textfield where the user can type keywords to search for books
   * @returns {ReactElement}
   * @private
   */
  render() {
    const { query } = this.state;
    return (
      <div className="search-books-input-wrapper">
        <input
          type="text"
          placeholder="Search by title or author"
          value={query}
          onChange={this.onTyping}
        />
      </div>
    );
  }
}

export default SearchBar;
