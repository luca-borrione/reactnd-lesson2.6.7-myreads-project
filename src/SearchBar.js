import React from 'react'
import PropTypes from 'prop-types'
import * as BooksAPI from './BooksAPI'
import debounce from 'lodash.debounce'

/**
 * @class BooksGrid
 * @extends React.Component
 * @classdesc
 * Prints a list of books
 * @hideconstructor
 */
class SearchBar extends React.Component {

	static SEARCH_LIMIT = 20;
	static DEBOUNCING_TIME = 250; // milliseconds

	/**
	 * @property {Object} propTypes - Intended types passed to the component
	 * @property {TBook[]} propTypes.books - List of the books to show
	 * @property {function} propTypes.updateBookShelf
	 * 	Method of the [App]{@link App#updateBookShelf} component for updating the list of books
	 * 	in the shelves
	 * @static
	 */
	static propTypes = {
		showResult: PropTypes.func.isRequired
	};

	state = {
		keywords: ''
	};

	async = {
		onTyping: null
	};

	constructor(props) {
		super(props);
		const { DEBOUNCING_TIME } = this.constructor;

		this.onTyping = this.onTyping.bind(this);
		this.search = debounce(this.search, DEBOUNCING_TIME);
	}

	/**
	 * @description
	 * Performs a search to find books matching the query string typed by the user.
	 * If no books are found, or in case of search error, the books list will be cleared.
	 * If books are returned by the search, then they will be listed undernit the search textfield.
	 * @param {string} query
	 * @returns {void}
	 */
	async search(query, resolve, reject) {
		const { SEARCH_LIMIT } = this.constructor;
		const { getBookShelf } = this.props;

		let books;

		if (query === '') {
			books = [];
			return resolve ? resolve(books) : books;
		}

		try {

			const fetchedBooks = await BooksAPI.search(query, SEARCH_LIMIT);
			if (!fetchedBooks || fetchedBooks.error) {
				books = [];
			} else {
				books = fetchedBooks.map( book => {
					// The book literal objects returned by the real search do not contain the shelf property,
					// so we need to define it by ourselves.
					// The ones returned by the mock search alredy contain it, so we will skip this in that case.
					if (!book.shelf) {
						book.shelf = getBookShelf(book.id);
					}
					return book;
				});
			}

			return resolve ? resolve(books) : books;

		} catch (error) {

			if (reject) {
				reject(error);
			} else {
				throw new Error(error);
			}
		}
	}

	fetchBooks(keywords) {
		return new Promise( async (resolve, reject) => {
				this.search(keywords, resolve, reject);
			}).then( async books => {
				books.status = 'ready';
				return books;
			}).catch( error => {
				console.error('ERROR: ', error);
				const books = [];
				books.status = 'error';
				return books;
			});
	}

	/**
	 * Updates the internal state according to the keywords typed by the user,<br>
	 * then it clears the books list if the textfield is empty,<br>
	 * otherwise it performs a search based on the typed keywords.
	 * @param {SyntheticEvent} event
	 * @returns {void}
	 */
	onTyping(event) {
		const keywords = event.target.value;
		this.async.onTyping = new Promise( async resolve => {
			await this.setState({ keywords });
			const books = await this.fetchBooks(keywords);
			await this.props.showResult(books);
			resolve(books);
		});
	}


	shouldComponentUpdate(nextProps, nextState) {
		return this.state.keywords !== nextState.keywords;
	}

	/**
	 * @description
	 * Shows the given list of books by rendering the [Book]{@link Book} component
	 * @returns {ReactElement}
	 */
	render() {
		const { keywords } = this.state;
		console.log('>> BAR RENDERED <<', keywords);
		return (
			<div className="search-books-input-wrapper">
				<input type="text" placeholder="Search by title or author"
					value={keywords}
					onChange={this.onTyping} />
			</div>
		);
	}
}


export default SearchBar;
