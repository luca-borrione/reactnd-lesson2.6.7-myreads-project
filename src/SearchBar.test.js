import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import * as BooksAPI from './BooksAPI'; // mocked
import SearchBar from './SearchBar';
import { mount, shallow } from 'enzyme';
import debounce from 'lodash.debounce';

describe('SearchBar', () => {

	beforeEach(() => {
		jest.resetModules();
		BooksAPI.__resetBooks();
	});

	afterEach(() => {
		jest.clearAllMocks();
		jest.restoreAllMocks();
	});

	const getBookShelf = () => {};
	const showResult = () => {};


	it('renders without crashing', () => {
		const div = document.createElement('div');
		ReactDOM.render(
			<SearchBar
				getBookShelf={getBookShelf}
				showResult={showResult} />, div);
		ReactDOM.unmountComponentAtNode(div);
	});


	it('renders correctly', () => {
		const tree = renderer.create(
			<SearchBar
				getBookShelf={getBookShelf}
				showResult={showResult} />,
		)
		.toJSON();

		expect(tree).toMatchSnapshot();
	});


	it('contains a debounced search method', () => {
		const api = jest.spyOn(BooksAPI, 'search');
		const wrapper = shallow(
			<SearchBar
				getBookShelf={getBookShelf}
				showResult={showResult} />
		);
		const component = wrapper.instance();

		const query = 't';
		component.search(query);
		expect(debounce).toHaveBeenCalledTimes(1);
		expect(api).toHaveBeenCalledTimes(1);
	});


	it("fetches books when typing keywords in the search field and will pass them to prop showResult", async () => {
		const query = 't';
		const showResult = jest.fn();

		const mounted = mount(
			<SearchBar
				getBookShelf={getBookShelf}
				showResult={showResult} />
		);
		const component = mounted.instance();

		const input = mounted.find('input');
		input.simulate('change', {
			target: { value : query}
		});
		const books = await component.async.onTyping;

		expect(component.state.keywords).toBe(query);
		expect(showResult).toHaveBeenCalledTimes(1);
		expect(books).toHaveLength(5);
		expect(showResult).toHaveBeenCalledWith(books);
	});
});
