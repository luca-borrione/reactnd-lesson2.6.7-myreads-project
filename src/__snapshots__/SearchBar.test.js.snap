// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar renders correctly 1`] = `
<div
  className="search-books-input-wrapper"
>
  <input
    onChange={[Function]}
    placeholder="Search by title or author"
    type="text"
    value=""
  />
</div>
`;
